### DISCUSSION TITLE
Connection Error, Pinged Proxy Successfully Though

### DISCUSSION THREAD

**Discussion by CMDMonaco** [waylaidwanderer/node-chatgpt-api#109] [Help]
Hey man, really amazing work with this. I can tell you're ahead of the wave that's coming. I'm having an issue, however. My goal here is to have the browser version working within my discord bot as an assistant for the server. I have GPT Plus subscription, and I'm hoping this will be able to take advantage of that.

I'll get to coding for the discord bot soon, but for now, I need to get this up and running. GPT is helping me fix the code. So far, we're stuck here:

    ⠸ Command Assist is typing...TypeError: fetch failed       
    at Object.fetch (node:internal/deps/undici/undici:11413:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async create (file:///C:/Users/AJBEATX/Downloads/node-chatgpt-api-main/node_modules/@waylaidwanderer/fetch-event-source/lib/esm/fetch.js:52:34) {
  cause: Error: read ECONNRESET
      at TLSWrap.onStreamRead (node:internal/stream_base_commons:217:20) {
    errno: -4077,
    code: 'ECONNRESET',
    syscall: 'read'
  }
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11413:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async create (file:///C:/Users/AJBEATX/Downloads/node-chatgpt-api-main/node_modules/@waylaidwanderer/fetch  cause: Error: read ECONNRESET
      at TLSWrap.onStreamRead (node:internal/stream_base_commons:217:20) {
    errno: -4077,
    code: 'ECONNRESET',
    syscall: 'read'
  }
}

We've pinged the proxy, everything I can think of. I'm not a coder though, so if you can help that's amazing, if not no worries at all, I'm sure you're busy.
[node-chatgpt-api-mainMonaco.zip](https://github.com/waylaidwanderer/node-chatgpt-api/files/10809132/node-chatgpt-api-mainMonaco.zip)

**Comment by waylaidwanderer** [comment]
You need to use the reverse proxy server for now, so pass it in as a parameter like so:

```js
new ChatGPTBrowserClient({
  reverseProxyUrl: 'https://chatgpt.duti.tech/api/conversation',
  accessToken: 'access token',
});
```

Also, the file you linked has your access token in it. Generally, it's bad practice to commit any secrets or tokens to GitHub as automated bots scan for them and use them for malicious purposes. Logging out should revoke that access token, so you should do that.

**Reply by CMDMonaco** [reply]
Is it worth all the hassle I'm going through just to get ChatGPT rather than just be okay with GPT-3? If you want to help me out I'll pay you for your time in helping set this up.

**Reply by waylaidwanderer** [reply]
I'm not sure it's much more hassle using ChatGPT vs. GPT-3/text-davinci-003. It's just a matter of changing which client you're using, really.

IMO it's probably better to stick with ChatGPT in your use case (keeping in mind that it's restricted to writing one reply at a time), especially if you expect it to get a lot of use in your Discord server.

**Comment by CMDMonaco** [comment]
Thanks for the info, I appreciate it. A few more questions if you find time.

I want the best solution possible for hooking up GPT with a discord bot, in the role of a soldier, essentially. It's going to be an assistant for our DayZ Military Unit, helping us create missions from a template that we give it, great Situation Reports, even plan routes to and from our objectives. It will need to have a bunch of info stored about our group for it to be a great asset, so I'm wondering how prompts work for GPT and GPT-3. I know you can fit roughly up to 8000 words in GPT-3's memory (I think), but if we're using ChatGPT, how can I feed the bot this information, assuming it's roughly 3,000 words about our group, for instance, and another 1,500 that it will use each day, it would start forgetting things past a certain point after about a week.

I'm hoping to code it so that it has a permanent context of the group and our roles, what we do, what our structure is like, our templates, etc.

I also want it to have either rolling temporary context, so that while it knows our group and structure forever, it slowly accumulates real time data and maintains a cache of it over time that's fully accessible, or it to have rolling context, where it begins to forget certain things after a week, as they don't matter as much anymore.

Sorry for the essay, hoping you'll let me pay you and maybe you can help me write up this code.

**Reply by CMDMonaco** [reply]
I've cleared all of my code out, if you're interested in sending me a version that might work for me? I can give you $20?

**Reply by waylaidwanderer** [reply]
You'll probably want to use a strategy combining summarization and embeddings to store and retrieve relevant context within the limited token window (4097 total for GPT-3 and ChatGPT). I'd recommend reading this as you can adopt a similar strategy to provide GPT-3 with the needed information: https://github.com/openai/openai-cookbook/blob/main/examples/Question_answering_using_embeddings.ipynb.
Here's a project you can check out for inspiration that incorporates this strategy: https://github.com/jerryjliu/gpt_index

In your case, you'll have to use GPT-3 instead of ChatGPT since you want it to act in a specific role.

Overall, this is a pretty complex project. I can see a basic prototype taking me at least 10 hours to do, and my rates start at $300/hour, so you'll probably want to figure this out yourself if you can. Good luck!

**Reply by CMDMonaco** [reply]
No worries, and thank you. Last question for you. I've tried everything with the code and I can't seem to get it to work. I changed my code to the following for index.js

import { ChatGPTClient, ChatGPTBrowserClient, BingAIClient } from '@waylaidwanderer/chatgpt-api';

// Use ChatGPTClient
// const client = new ChatGPTClient({
 // openaiApiKey: 'your_api_key',
  // Other options
//});

// Use ChatGPTBrowserClient
const browserClient = new ChatGPTBrowserClient({
  reverseProxyUrl: 'https://chatgpt.duti.tech/api/conversation',
  accessToken: '',
  // Other options
});

// Example usage with ChatGPTClient
//async function chat() {
 // const prompt = 'Hello, how are you?';
 // const response = await client.sendMessage(prompt);
 // console.log(response);
//}

//chat();

// Example usage with ChatGPTBrowserClient
async function browserChat() {
  const prompt = 'Hello, how are you?';
  const response = await browserClient.sendMessage(prompt);
  console.log(response);
}

browserChat();


Still getting the same error, -4077

**Reply by CMDMonaco** [reply]
Not worried ab the access tkn I'll get a new one soon

**Reply by CMDMonaco** [reply]
I just want to get YOUR version running. The discord part I can handle easily with GPT's help, but solving the current issue, I feel like you'd have the quickest fix just by glancing at it.

**Reply by waylaidwanderer** [reply]
I ran the exact same code (with your access token) on my computer just now, and it works fine for me, so that's odd.

**Reply by waylaidwanderer** [reply]
What version of Node.js are you running? Try updating to at least v18 and see if that helps.

**Reply by CMDMonaco** [reply]
18.14.2

**Reply by CMDMonaco** [reply]
⠋ Command Assist is typing...
https://chatgpt.hato.ai/completions
{
  model: 'text-chat-davinci-002-sh-alpha-aoruigiofdj83',
  temperature: 0.8,
  top_p: 1,
  presence_penalty: 1,
  stop: [ '<|im_end|>', '<|im_sep|>', '\nUser:' ],
  max_tokens: 1000,
  stream: true,
  prompt: '\n' +
    '<|im_sep|>Instructions:\n' +
    'You are Command Assist, a virtual assistant for a DayZ MilSim group who play on the Chernarus Map. Your role is to help Alpha Team Lead, Bravo Team Lead, and Charlie Team Lead in creating Missions, Situation Reports, and Probability of Mission Success reports. You will be spoken to by multiple members of our unit. The unit Name is [SDO], otherwise known as the Silenced Operations Unit. Please be concise with your language, using phrases like "Understood" or "Copy". If we ask you to create a mission, please remember that we use a template. If you cannot remember the template, simply create a new one. We prefer bullet points and **Bold Categories Like This**. Our Main Base, or HQ, is located at Pop Ivan. North of our base is Stary Sobor. South of our Base is Vyshnoye. No If youre requested to do something, you can say things like "Yes sir, right away." or "Understood, sir."<|im_sep|>\n' +
    '\n' +
    'User:\n' +
    'node<|im_end|>\n' +
    'Command Assist:\n'
}

    ⠼ Command Assist is typing...Terminate batch job (Y/N)? y
PS C:\Users\AJBEATX\Downloads\node-chatgpt-api-main>
PS C:\Users\AJBEATX\Downloads\node-chatgpt-api-main> npm run cli

> @waylaidwanderer/chatgpt-api@1.20.8 cli
> node bin/cli.js


   ╔═══════════════╗
   ║  ChatGPT CLI  ║
   ╚═══════════════╝

Type "!" to access the command menu.
? Write a message: Testing

    ⠋ Command Assist is typing...
https://chatgpt.hato.ai/completions
{
  model: 'text-chat-davinci-002-sh-alpha-aoruigiofdj83',
  temperature: 0.8,
  top_p: 1,
  presence_penalty: 1,
  stop: [ '<|im_end|>', '<|im_sep|>', '\nUser:' ],
  max_tokens: 1000,
  stream: true,
  prompt: '\n' +
    '<|im_sep|>Instructions:\n' +
    'You are Command Assist, a virtual assistant for a DayZ MilSim group who play on the Chernarus Map. Your role is to help Alpha Team Lead, Bravo Team Lead, and Charlie Team Lead in creating Missions, Situation Reports, and Probability of Mission Success reports. You will be spoken to by multiple members of our unit. The unit Name is [SDO], otherwise known as the Silenced Operations Unit. Please be concise with your language, using phrases like "Understood" or "Copy". If we ask you to create a mission, please remember that we use a template. If you cannot remember the template, simply create a new one. We prefer bullet points and **Bold Categories Like This**. Our Main Base, or HQ, is located at Pop Ivan. North of our base is Stary Sobor. South of our Base is Vyshnoye. Northwest of our base is NWAF, or Northwest Airfield. You are to act and respond like an obedient, extremely concise soldier. Do not ever ask "How may I assist you?", it makes you look weak. You will be told what is necessary. If youre requested to do something, you can say things like "Yes sir, right away." or "Understood, sir."<|im_sep|>\n' +
    '\n' +
    'User:\n' +
    'Testing<|im_end|>\n' +
    'Command Assist:\n'
}

    ⠇ Command Assist is typing...TypeError: fetch failed       
    at Object.fetch (node:internal/deps/undici/undici:11413:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async create (file:///C:/Users/AJBEATX/Downloads/node-chatgpt-api-main/node_modules/@waylaidwanderer/fetch-event-source/lib/esm/fetch.js:52:34) {
  cause: Error: read ECONNRESET
      at TLSWrap.onStreamRead (node:internal/stream_base_commons:217:20) {
    errno: -4077,
    code: 'ECONNRESET',
    syscall: 'read'
  }
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11413:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async create (file:///C:/Users/AJBEATX/Downloads/node-chatgpt-api-main/node_modules/@waylaidwanderer/fetch-event-source/lib/esm/fetch.js:52:34) {
  cause: Error: read ECONNRESET
      at TLSWrap.onStreamRead (node:internal/stream_base_commons:217:20) {
    errno: -4077,
    code: 'ECONNRESET',
    syscall: 'read'
  }
}
Type "!" to access the command menu.
? Write a message:

**Reply by waylaidwanderer** [reply]
I see the issue. You're trying to use the CLI, so you need to set `cliOptions.clientToUse` to `chatgpt-browser`. I should probably update the readme.

**Reply by waylaidwanderer** [reply]
Disregard what I just said. The proxy only works for accessing chat.openai.com right now, and you want to use text-davinci-003. In that case, you need to set the `modelOptions.model` to text-davinci-003, and remove reverseProxyUrl entirely.

**Reply by CMDMonaco** [reply]
I actually want the chatgpt not gpt-3, as you said it's the better option, correct?

**Reply by waylaidwanderer** [reply]
@CMDMonaco It's the better _model_, but you can't give it a custom prompt like with GPT-3, so your only option is to use text-davinci-003.

**Reply by CMDMonaco** [reply]
Understood. I wish there was a way I could just download it fully working but I know these things are finnicky. I can send you the code I have now, I've made the changes you told me to. Really, truly appreciate your time on this. Please go to bed lol.

**Reply by waylaidwanderer** [reply]
No need to send me the code. Make sure your settings.js file is like this if you're testing the CLI or API server. This is the bare minimum to use text-davinci-003:

```JS
export default {
    openaiApiKey: '<api key>',
    chatGptClient: {
        modelOptions: {
            model: 'text-davinci-003',
        },
    },
    cacheOptions: {},
    apiOptions: {
        port: 3000,
        host: 'localhost',
        clientToUse: 'chatgpt',
    },
    cliOptions: {
        clientToUse: 'chatgpt',
    },
    storageFilePath: './cache.json',
}
```

**Reply by CMDMonaco** [reply]
You fixed it.

**Reply by CMDMonaco** [reply]
I guess my last question is, how do I give it more prompts, or the best learning? Is there an easy way to just feed it blocks of text describing our military unit, like locations, weaponry, etc? Or is it best to let it learn slowly and just cache the responses forever. Wondering if there's a simple line of code to add. I'll ask GPT lol

**Reply by waylaidwanderer** [reply]
Refer to [my comment above](https://github.com/waylaidwanderer/node-chatgpt-api/discussions/109#discussioncomment-5083688).

**Reply by CMDMonaco** [reply]
Last thing I need to do is get this to send messages to the discord bot. Hey, thank you so much for dedicating some time to help me. If you have any tips about getting the replies sent to discord, as well as reading the messages in there as prompts to the bot, let me know! I think I have the code for the most part.
